wpilibc/src/main/native/cpp/Tracer.cpp

```diff
@@ -4,14 +4,226 @@
 
 #include "frc/Tracer.h"
 
+#include <memory>
+#include <string>
+#include <thread>
+#include <utility>
+#include <vector>
+
 #include <fmt/format.h>
+#include <networktables/DoubleTopic.h>
+#include <networktables/NetworkTable.h>
+#include <networktables/NetworkTableInstance.h>
+#include <units/time.h>
 #include <wpi/SmallString.h>
+#include <wpi/StringMap.h>
+#include <wpi/deprecated.h>
 #include <wpi/raw_ostream.h>
 
 #include "frc/Errors.h"
+#include "frc/RobotController.h"
 
 using namespace frc;
 
+std::atomic<bool> singleThreadedMode = false;
+std::atomic<bool> anyTracesStarted = false;
+
+class TracerState {
+ public:
+  TracerState() {
+    if (singleThreadedMode && anyTracesStarted) {
+      FRC_ReportWarning("Cannot start a new trace in single-threaded mode");
+      m_disabled = true;
+    }
+    anyTracesStarted = true;
+    auto inst = nt::NetworkTableInstance::GetDefault();
+    m_rootTable = inst.GetTable(
+        fmt::format("/Tracer/{}",
+                    std::hash<std::thread::id>{}(std::this_thread::get_id())));
+  }
+
+  std::string BuildStack() {
+    std::string stack = "";
+    for (size_t i = 0; i < m_traceStack.size(); ++i) {
+      stack += m_traceStack[i];
+      if (i < m_traceStack.size() - 1) {
+        stack += "/";
+      }
+    }
+    return stack;
+  }
+
+  std::string AppendTraceStack(std::string_view trace) {
+    m_stackSize++;
+    if (m_disabled) {
+      return "";
+    }
+    m_traceStack.push_back(trace);
+    return BuildStack();
+  }
+
+  std::string PopTraceStack() {
+    m_stackSize > 0 ? m_stackSize-- : m_stackSize;
+    if (m_disabled) {
+      return "";
+    }
+    if (m_traceStack.empty() || m_cyclePoisened) {
+      m_cyclePoisened = true;
+      return "";
+    }
+    std::string stack = BuildStack();
+    m_traceStack.pop_back();
+    return stack;
+  }
+
+  void EndCycle() {
+    if (m_disabled != m_disableNextCycle || m_cyclePoisened) {
+      // Gives publishers empty times,
+      // reporting no data is better than bad data
+      for (auto&& [_, publisher] : m_publishers) {
+        publisher.Set(0.0);
+      }
+      return;
+    } else {
+      // Update times for all already existing publishers
+      for (auto&& [key, publisher] : m_publishers) {
+        if (auto time = m_traceTimes.find(key); time != m_traceTimes.end()) {
+          publisher.Set(time->second.value());
+          m_traceTimes.erase(time);
+        } else {
+          publisher.Set(0.0);
+        }
+      }
+
+      // Create publishers for all new entries, add them to the heap
+      // and set their times
+      for (auto&& traceTime : m_traceTimes) {
+        auto topic = m_rootTable->GetDoubleTopic(traceTime.first());
+        if (auto publisher = topic.Publish(); publisher) {
+          publisher.Set(traceTime.second.value());
+          m_publishers.push_back(
+              std::make_pair(traceTime.first(), std::move(publisher)));
```

Comment:
Can this be replaced with `emplace_back(traceTime.first(), std::move(publisher)`?